services.py

```diff
@@ -88,42 +76,80 @@ def get_param_from_config(self, config_path: str, param_name: str):
+                    # который также работает
+                    parts = data.split(",")
+                    print(parts)
+                    if parts[2] == 'V':
+                        # V = Предупреждение, скорее всего, спутников нет в поле зрения ...
+                        print ("GPS receiver warning")
+                    else:
+                        # Получить данные о местоположении, которые были переданы с сообщением GPRMC. 
+                        # В этом примере интересуют только долгота и широта. Для других значений можно
+                        # обратиться к http://aprs.gids.nl/nmea/#rmc
+                        longitude = self.formatDegreesMinutes(parts[5], 3)
+                        latitude = self.formatDegreesMinutes(parts[3], 2)
+                        self.logger.info("Device position: lon = " + str(longitude) + ", lat = " + str(latitude))
+                else:
+                    # Обрабатывать другие сообщения NMEA и неподдерживаемые строки
+                    pass
+
+                gps.close()
+                return [latitude, longitude]
+        except Exception as e:
+            self.logger.error(f'Error getting GPS coordinates through serial port: {e}\n Falling back to IP geolocation...')
             try:
-                os.remove(path)
-                self.logger.info(f'Deleted file: {path}')
-            except PermissionError:
-                self.logger.error(f'Error deleting file: {path}')
-
-
-    def get_logger(self, name):
-        log_format = '%(asctime)s  %(name)8s  %(levelname)5s  %(message)s'
-        logging.basicConfig(level=logging.DEBUG,
-                            format=log_format,
-                            filename='dev.log',
-                            filemode='w')
-        return logging.getLogger(name)
-
-
-    def get_city_from_coordinates(self, latitude, longitude):
-        geoLoc = Nominatim(user_agent="GetLoc")
-
-        loc = self.get_lat_lon()
-        locname = geoLoc.reverse(f'50.0 40.0')
-        address = locname.raw['address']
-        print(address)
-
-    # latitude, longitude = get_current_location()
-    # if latitude and longitude:
-    #     city = get_city_from_coordinates(latitude, longitude)
-    #     if city:
-    #         print(f"The current city is: {city}")
-    #     else:
-    #         print("Could not determine the city.")
-    # else:
-    #     print("Could not determine the current location.")
+                g = geocoder.ip('me')
+                return g.latlng
+            except Exception as e:
+                self.logger.critical(f'Error getting GPS coordinates through IP geolocation: {e}')
+                raise RuntimeError(f'Error getting GPS coordinates through IP geolocation') from e
+
+    def save_json(self, dictionary, path: str):
+        with open(path, "w") as outfile:
+            json.dump(dictionary, outfile, indent=4)
+
+    def get_current_city(self):
```

Comment:
This method is not trivial, so you generally don't want to call it repeatedly unless you need to. So, you either want to make sure it is only called once, or you calculate and save it as a property during `__init__()`, or you memoize it, say, using the `@functools.cached_property` decorator (then you will rename this method to `current_city` and access it as a property - it will only be called once per the `Services` instance, which you should only have one of)